// @wpackio/scripts/lib/config/fileLoader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getFileLoaderOptions = getFileLoaderOptions;
exports.issuerForStyleFiles = issuerForStyleFiles;
exports.issuerForNonStyleFiles = issuerForNonStyleFiles;
exports.issuerForJsTsFiles = issuerForJsTsFiles;
exports.issuerForNonJsTsFiles = issuerForNonJsTsFiles;
exports.getFileLoaderForJsAndStyleAssets = getFileLoaderForJsAndStyleAssets;

/**
 * This file holds useful utility functions for file-loader
 */

/**
 * Get options for file-loader.
 *
 * This takes into account the appDir, development or production mode and
 * publicPath for file-loader usage from css files.
 *
 * @param appDir Application directory (corresponds to files[i].name).
 * @param isDev Whether this is for development or production build.
 * @param publicPath Whether to include publicPath.
 */
function getFileLoaderOptions(appDir, isDev, publicPath) {
  const fileLoaderOptions = {
    name: `[name]-[hash:8].[ext]`,
    outputPath: `${appDir}/assets/`
  };

  if (publicPath) {
    // Here mention the public path relative to the css
    // file directory, but only during production mode
    // for development mode, it is still undefined
    fileLoaderOptions.publicPath = isDev ? undefined : `assets/`;
  }

  return fileLoaderOptions;
}
/**
 * Webpack module.rules[].issuer function to check if the issuer is
 * a style file (scss, sass, css).
 */


function issuerForStyleFiles(location) {
  return /\.(sa|sc|c)ss$/.test(location);
}
/**
 * Webpack module.rules[].issuer function to check if the issuer is NOT
 * a style file (scss, sass, css).
 */


function issuerForNonStyleFiles(location) {
  return !/\.(sa|sc|c)ss$/.test(location);
}
/**
 * Webpack module.rules[].issuer function to check if the issuer is NOT a
 * JavaScript (js,jsx) or TypeScript(ts,tsx) file.
 */


function issuerForJsTsFiles(location) {
  return /\.(j|t)sx?$/.test(location);
}
/**
 * Webpack module.rules[].issuer function to check if the issuer is a JavaScript
 * (js,jsx) or TypeScript(ts,tsx) file.
 */


function issuerForNonJsTsFiles(location) {
  return !/\.(j|t)sx?$/.test(location);
}
/**
 * Get file-loader with proper options set for javascript and style files.
 *
 * It makes sure that assets are put into proper directory and relativePath
 * works fine for css imports.
 *
 * @param appDir Application directory (corresponds to files[i].name).
 * @param isDev Whether this is for development or production build.
 * @param loader Name of the loader, defaults to `file-loader`.
 */


function getFileLoaderForJsAndStyleAssets(appDir, isDev, loader = 'file-loader') {
  // create file rules
  // But use relativePath for style type resources like sass, scss or css
  // This is needed because we can't know the absolute publicPath
  // of CSS imported assets.
  const fileLoaderTest = /\.(woff|woff2|eot|ttf|otf|png|jpg|gif)(\?v=\d+\.\d+\.\d+)?$/;
  const fileRulesNonStyle = {
    test: fileLoaderTest,
    use: [{
      loader,
      options: getFileLoaderOptions(appDir, isDev, false)
    }],
    issuer: issuerForNonStyleFiles
  };
  const fileRulesStyle = {
    test: fileLoaderTest,
    use: [{
      loader,
      options: getFileLoaderOptions(appDir, isDev, true)
    }],
    issuer: issuerForStyleFiles
  };
  return {
    fileRulesNonStyle,
    fileRulesStyle
  };
}
